File: apps/lambda_ts/scripts/deploy.ts
import { execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const BUILD_DIR = path.join(__dirname, '../.build');
const AWS_REGION = process.env.AWS_REGION || 'us-east-1';

interface DeployConfig {
  functionName: string;
  handler: string;
  runtime: string;
  role: string;
  timeout?: number;
  memorySize?: number;
  environment?: Record<string, string>;
}

async function deployFunction(name: string): Promise<void> {
  const buildPath = path.join(BUILD_DIR, name);
  const handlerPath = path.join(buildPath, 'index.mjs');

  if (!fs.existsSync(handlerPath)) {
    console.error(
      `❌ Build not found for ${name}. Run 'pnpm build:${name}' first.`
    );
    process.exit(1);
  }

  // Load deployment config if it exists
  const configPath = path.join(__dirname, '../deploy-config.json');
  let config: Record<string, DeployConfig> = {};

  if (fs.existsSync(configPath)) {
    config = JSON.parse(fs.readFileSync(configPath, 'utf-8'));
  }

  const functionConfig = config[name] || {
    functionName: `guandan-${name}`,
    handler: 'index.handler',
    runtime: 'nodejs20.x',
    role:
      process.env.LAMBDA_ROLE_ARN ||
      'arn:aws:iam::ACCOUNT_ID:role/lambda-execution-role',
    timeout: 30,
    memorySize: 256,
  };

  console.log(`🚀 Deploying ${name} to AWS Lambda...`);
  console.log(`   Function name: ${functionConfig.functionName}`);
  console.log(`   Region: ${AWS_REGION}`);

  // Create a zip file
  const zipPath = path.join(BUILD_DIR, `${name}.zip`);

  try {
    // Create zip
    execSync(`cd ${buildPath} && zip -r ${zipPath} .`, { stdio: 'inherit' });

    // Check if function exists
    let functionExists = false;
    try {
      execSync(
        `aws lambda get-function --function-name ${functionConfig.functionName} --region ${AWS_REGION}`,
        { stdio: 'pipe' }
      );
      functionExists = true;
    } catch {
      functionExists = false;
    }

    if (functionExists) {
      // Update existing function
      console.log('   Updating existing function...');
      execSync(
        `aws lambda update-function-code --function-name ${functionConfig.functionName} --zip-file fileb://${zipPath} --region ${AWS_REGION}`,
        { stdio: 'inherit' }
      );
    } else {
      // Create new function
      console.log('   Creating new function...');
      const envVars = functionConfig.environment
        ? `--environment Variables={${Object.entries(functionConfig.environment)
            .map(([k, v]) => `${k}=${v}`)
            .join(',')}}`
        : '';

      execSync(
        `aws lambda create-function \
          --function-name ${functionConfig.functionName} \
          --runtime ${functionConfig.runtime} \
          --role ${functionConfig.role} \
          --handler ${functionConfig.handler} \
          --zip-file fileb://${zipPath} \
          --timeout ${functionConfig.timeout} \
          --memory-size ${functionConfig.memorySize} \
          --region ${AWS_REGION} \
          ${envVars}`,
        { stdio: 'inherit' }
      );
    }

    console.log(`✅ Deployed ${name} successfully`);
  } catch (error) {
    console.error(`❌ Failed to deploy ${name}:`, error);
    process.exit(1);
  }
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.error('❌ Please specify a function name to deploy');
    console.error('   Usage: pnpm deploy <function-name>');
    process.exit(1);
  }

  const functionName = args[0];
  await deployFunction(functionName);
}

main().catch((error) => {
  console.error('Deployment failed:', error);
  process.exit(1);
});

File: apps/lambda_py/scripts/deploy.py
#!/usr/bin/env python3
"""Deployment script for Lambda functions."""

import argparse
import json
import os
import shutil
import subprocess
import sys
from pathlib import Path


def get_project_root() -> Path:
    """Get the project root directory."""
    return Path(__file__).parent.parent


def get_build_dir() -> Path:
    """Get the build directory."""
    return get_project_root() / ".build"


def get_deploy_config() -> dict:
    """Load deployment configuration."""
    config_file = get_project_root() / "deploy-config.json"
    
    if not config_file.exists():
        return {}
    
    with open(config_file) as f:
        return json.load(f)


def create_zip(function_name: str) -> Path:
    """
    Create a zip file for deployment.
    
    Args:
        function_name: Name of the function
        
    Returns:
        Path to the zip file
    """
    build_path = get_build_dir() / function_name
    
    if not build_path.exists():
        print(f"❌ Build not found for {function_name}. Run 'npm run build:{function_name}' first.")
        sys.exit(1)
    
    zip_path = get_build_dir() / f"{function_name}.zip"
    
    # Remove existing zip
    if zip_path.exists():
        zip_path.unlink()
    
    # Create zip
    print(f"📦 Creating deployment package...")
    shutil.make_archive(
        str(zip_path.with_suffix("")),
        "zip",
        str(build_path),
    )
    
    return zip_path


def function_exists(function_name: str, region: str) -> bool:
    """
    Check if a Lambda function exists.
    
    Args:
        function_name: AWS Lambda function name
        region: AWS region
        
    Returns:
        True if function exists, False otherwise
    """
    try:
        subprocess.run(
            [
                "aws",
                "lambda",
                "get-function",
                "--function-name",
                function_name,
                "--region",
                region,
            ],
            check=True,
            capture_output=True,
        )
        return True
    except subprocess.CalledProcessError:
        return False


def deploy_function(function_name: str) -> None:
    """
    Deploy a Lambda function to AWS.
    
    Args:
        function_name: Name of the function to deploy
    """
    config = get_deploy_config()
    region = os.environ.get("AWS_REGION", "us-east-1")
    
    # Get function config or use defaults
    func_config = config.get(function_name, {})
    aws_function_name = func_config.get("functionName", f"guandan-{function_name}")
    handler = func_config.get("handler", f"functions.{function_name}.handler.handler")
    runtime = func_config.get("runtime", "python3.13")
    role = func_config.get("role", os.environ.get("LAMBDA_ROLE_ARN", ""))
    timeout = func_config.get("timeout", 30)
    memory_size = func_config.get("memorySize", 256)
    environment = func_config.get("environment", {})
    
    if not role:
        print("❌ Lambda execution role not specified.")
        print("   Set LAMBDA_ROLE_ARN environment variable or add 'role' to deploy-config.json")
        sys.exit(1)
    
    print(f"🚀 Deploying {function_name} to AWS Lambda...")
    print(f"   Function name: {aws_function_name}")
    print(f"   Region: {region}")
    
    # Create zip
    zip_path = create_zip(function_name)
    
    try:
        if function_exists(aws_function_name, region):
            # Update existing function
            print("   Updating existing function...")
            subprocess.run(
                [
                    "aws",
                    "lambda",
                    "update-function-code",
                    "--function-name",
                    aws_function_name,
                    "--zip-file",
                    f"fileb://{zip_path}",
                    "--region",
                    region,
                ],
                check=True,
            )
            
            # Update configuration
            cmd = [
                "aws",
                "lambda",
                "update-function-configuration",
                "--function-name",
                aws_function_name,
                "--handler",
                handler,
                "--timeout",
                str(timeout),
                "--memory-size",
                str(memory_size),
                "--region",
                region,
            ]
            
            if environment:
                env_vars = ",".join(f"{k}={v}" for k, v in environment.items())
                cmd.extend(["--environment", f"Variables={{{env_vars}}}"])
            
            subprocess.run(cmd, check=True)
        else:
            # Create new function
            print("   Creating new function...")
            cmd = [
                "aws",
                "lambda",
                "create-function",
                "--function-name",
                aws_function_name,
                "--runtime",
                runtime,
                "--role",
                role,
                "--handler",
                handler,
                "--zip-file",
                f"fileb://{zip_path}",
                "--timeout",
                str(timeout),
                "--memory-size",
                str(memory_size),
                "--region",
                region,
            ]
            
            if environment:
                env_vars = ",".join(f"{k}={v}" for k, v in environment.items())
                cmd.extend(["--environment", f"Variables={{{env_vars}}}"])
            
            subprocess.run(cmd, check=True)
        
        print(f"✅ Deployed {function_name} successfully")
    except subprocess.CalledProcessError as e:
        print(f"❌ Failed to deploy {function_name}: {e}")
        sys.exit(1)


def main() -> None:
    """Main entry point."""
    parser = argparse.ArgumentParser(description="Deploy Lambda functions to AWS")
    parser.add_argument(
        "function",
        help="Function name to deploy",
    )
    
    args = parser.parse_args()
    deploy_function(args.function)


if __name__ == "__main__":
    main()
